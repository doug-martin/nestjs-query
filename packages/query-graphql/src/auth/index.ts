export * from './authorizer';
export * from './default-crud.authorizer';
export * from './tokens';
