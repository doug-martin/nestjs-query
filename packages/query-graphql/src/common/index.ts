export { DTONamesOpts, getDTONames, DTONames, getDTOIdTypeOrDefault } from './dto.utils';
export * from './external.utils';
export * from './resolver.utils';
export * from './object.utils';
