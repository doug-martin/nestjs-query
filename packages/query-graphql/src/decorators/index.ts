export {
  FilterableField,
  FilterableFieldOptions,
  FilterableFieldDescriptor,
  getFilterableFields,
} from './filterable-field.decorator';
export { ResolverMethodOpts } from './resolver-method.decorator';
export {
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  Relation,
  FilterableRelation,
  RelationDecoratorOpts,
  RelationTypeFunc,
  getRelations,
} from './relation.decorator';
export * from './resolver-mutation.decorator';
export * from './resolver-query.decorator';
export * from './resolver-field.decorator';
export { Reference, ReferenceDecoratorOpts, ReferenceTypeFunc } from './reference.decorator';
export { ResolverSubscription, SubscriptionResolverMethodOpts } from './resolver-subscription.decorator';
export { InjectPubSub } from './inject-pub-sub.decorator';
export * from './skip-if.decorator';
export * from './aggregate-query-param.decorator';
export * from './hook.decorator';
export * from './decorator.utils';
export * from './hook-args.decorator';
export * from './authorizer.decorator';
export * from './inject-authorizer.decorator';
export * from './inject-custom-authorizer.decorator';
export * from './key-set.decorator';
export * from './authorize-filter.decorator';
export * from './query-options.decorator';
export * from './id-field.decorator';
