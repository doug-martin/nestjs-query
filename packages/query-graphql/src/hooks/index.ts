export * from './hooks';
export { createDefaultHook } from './default.hook';
export * from './tokens';
export * from './types';
