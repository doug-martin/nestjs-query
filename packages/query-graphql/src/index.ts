export * from './types';
export {
  FilterableField,
  FilterableFieldOptions,
  ResolverMethodOpts,
  Relation,
  FilterableRelation,
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  RelationTypeFunc,
  RelationDecoratorOpts,
  Reference,
  ReferenceTypeFunc,
  ReferenceDecoratorOpts,
  InjectPubSub,
  BeforeCreateOne,
  BeforeCreateMany,
  BeforeUpdateOne,
  BeforeUpdateMany,
  BeforeDeleteOne,
  BeforeDeleteMany,
  BeforeQueryMany,
  BeforeFindOne,
  InjectAuthorizer,
  InjectCustomAuthorizer,
  Authorize,
  AuthorizerFilter,
  RelationAuthorizerFilter,
  HookArgs,
  MutationHookArgs,
  KeySet,
  QueryOptions,
  IDField,
  IDFieldOptions,
} from './decorators';
export * from './resolvers';
export * from './federation';
export { DTONamesOpts } from './common';
export { NestjsQueryGraphQLModule } from './module';
export { AutoResolverOpts } from './providers';
export { pubSubToken, GraphQLPubSub } from './subscription';
export { Authorizer, CustomAuthorizer, AuthorizerOptions, AuthorizationContext, OperationGroup } from './auth';
export {
  Hook,
  HookTypes,
  BeforeCreateOneHook,
  BeforeCreateManyHook,
  BeforeUpdateOneHook,
  BeforeUpdateManyHook,
  BeforeDeleteOneHook,
  BeforeDeleteManyHook,
  BeforeQueryManyHook,
  BeforeFindOneHook,
} from './hooks';
export { AuthorizerInterceptor, AuthorizerContext, HookInterceptor, HookContext } from './interceptors';
