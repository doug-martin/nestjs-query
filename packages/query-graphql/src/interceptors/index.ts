export * from './hook.interceptor';
export * from './authorizer.interceptor';
