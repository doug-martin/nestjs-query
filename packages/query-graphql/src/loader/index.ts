export * from './find-relations.loader';
export * from './query-relations.loader';
export * from './count-relations.loader';
export * from './aggregate-relations.loader';
export * from './dataloader.factory';
