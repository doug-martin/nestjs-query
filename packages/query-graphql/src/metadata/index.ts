import { GraphQLQueryMetadataStorage } from './metadata-storage';

export { FilterableFieldDescriptor } from './metadata-storage';

/** @internal */
let storage: GraphQLQueryMetadataStorage;
/** @internal */
export function getMetadataStorage(): GraphQLQueryMetadataStorage {
  if (!storage) {
    storage = new GraphQLQueryMetadataStorage();
  }
  return storage;
}
