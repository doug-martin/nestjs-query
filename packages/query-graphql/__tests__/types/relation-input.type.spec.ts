// eslint-disable-next-line max-classes-per-file
import { plainToClass } from 'class-transformer';
import { validateSync } from 'class-validator';
import { InputType, Resolver, Query, Args, Int, ObjectType } from '@nestjs/graphql';
import { FilterableField, IDField, RelationInputType } from '../../src';
import { generateSchema } from '../__fixtures__';

describe('RelationInputType', (): void => {
  @ObjectType()
  class ParentDTO {
    @FilterableField()
    field!: string;
  }

  @ObjectType()
  class ParentCustomIDDTO {
    @IDField(() => String)
    id!: string;
  }

  @ObjectType()
  class RelationDTO {
    @FilterableField()
    relationField!: string;
  }

  @ObjectType()
  class RelationCustomIDDTO {
    @IDField(() => String)
    relationId!: string;
  }

  @InputType()
  class RelationInput extends RelationInputType(ParentDTO, RelationDTO) {}

  it('should create an input type with an id and relationId', async () => {
    @Resolver()
    class RelationInputTypeSpec {
      @Query(() => Int)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args('input') input: RelationInput): number {
        return 1;
      }
    }
    const schema = await generateSchema([RelationInputTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('should create an input type with a custom id for the parent', async () => {
    @InputType()
    class RelationCustomParentIdInput extends RelationInputType(ParentCustomIDDTO, RelationDTO) {}

    @Resolver()
    class RelationCustomIdInputTypeSpec {
      @Query(() => Int)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args('input') input: RelationCustomParentIdInput): number {
        return 1;
      }
    }
    const schema = await generateSchema([RelationCustomIdInputTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('should create an input type with a custom id for the relation', async () => {
    @InputType()
    class RelationCustomRelationIdInput extends RelationInputType(ParentDTO, RelationCustomIDDTO) {}

    @Resolver()
    class RelationCustomIdInputTypeSpec {
      @Query(() => Int)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args('input') input: RelationCustomRelationIdInput): number {
        return 1;
      }
    }
    const schema = await generateSchema([RelationCustomIdInputTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('should create an input type with a custom id for the parent and relation', async () => {
    @InputType()
    class RelationCustomParentAndRelationIdInput extends RelationInputType(ParentCustomIDDTO, RelationCustomIDDTO) {}

    @Resolver()
    class RelationCustomIdInputTypeSpec {
      @Query(() => Int)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args('input') input: RelationCustomParentAndRelationIdInput): number {
        return 1;
      }
    }
    const schema = await generateSchema([RelationCustomIdInputTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('should return the input when accessing the update field', () => {
    const input: RelationInputType = { id: 1, relationId: 2 };
    const it = plainToClass(RelationInput, input);
    expect(it.id).toEqual(input.id);
    expect(it.relationId).toEqual(input.relationId);
  });

  describe('validation', () => {
    it('should validate the id is defined', () => {
      const input = { relationId: 1 };
      const it = plainToClass(RelationInput, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [],
          constraints: {
            isNotEmpty: 'id should not be empty',
          },
          property: 'id',
          target: input,
        },
      ]);
    });

    it('should validate the id is not empty', () => {
      const input = { id: '', relationId: 1 };
      const it = plainToClass(RelationInput, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [],
          constraints: {
            isNotEmpty: 'id should not be empty',
          },
          property: 'id',
          target: input,
          value: '',
        },
      ]);
    });

    it('should validate that relationId is defined', () => {
      const input = { id: 1 };
      const it = plainToClass(RelationInput, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [],
          constraints: {
            isNotEmpty: 'relationId should not be empty',
          },
          property: 'relationId',
          target: input,
        },
      ]);
    });

    it('should validate that relationId is not empty', () => {
      const input: RelationInputType = { id: 1, relationId: '' };
      const it = plainToClass(RelationInput, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [],
          constraints: {
            isNotEmpty: 'relationId should not be empty',
          },
          property: 'relationId',
          target: input,
          value: input.relationId,
        },
      ]);
    });
  });
});
