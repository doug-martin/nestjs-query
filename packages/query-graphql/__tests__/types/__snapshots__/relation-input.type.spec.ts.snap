// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelationInputType should create an input type with a custom id for the parent 1`] = `
type Query {
  test(input: RelationCustomParentIdInput!): Int!
}

input RelationCustomParentIdInput {
  """The id of the record."""
  id: String!

  """The id of relation."""
  relationId: ID!
}

`;

exports[`RelationInputType should create an input type with a custom id for the parent and relation 1`] = `
type Query {
  test(input: RelationCustomParentAndRelationIdInput!): Int!
}

input RelationCustomParentAndRelationIdInput {
  """The id of the record."""
  id: String!

  """The id of relation."""
  relationId: String!
}

`;

exports[`RelationInputType should create an input type with a custom id for the relation 1`] = `
type Query {
  test(input: RelationCustomRelationIdInput!): Int!
}

input RelationCustomRelationIdInput {
  """The id of the record."""
  id: ID!

  """The id of relation."""
  relationId: String!
}

`;

exports[`RelationInputType should create an input type with an id and relationId 1`] = `
type Query {
  test(input: RelationInput!): Int!
}

input RelationInput {
  """The id of the record."""
  id: ID!

  """The id of relation."""
  relationId: ID!
}

`;
